src/hotspot/share/prims/scopedMemoryAccess.cpp

```diff
@@ -36,20 +38,35 @@
 #include "runtime/vframe.inline.hpp"
 
 template<typename Func>
-static bool for_scoped_method(JavaThread* jt, const Func& func) {
+static void for_scoped_methods(JavaThread* jt, const Func& func) {
   ResourceMark rm;
 #ifdef ASSERT
   LogMessage(foreign) msg;
   NonInterleavingLogStream ls{LogLevelType::Trace, msg};
   if (ls.is_enabled()) {
     ls.print_cr("Walking thread: %s", jt->name());
   }
+
+  bool would_have_bailed = false;
 #endif
 
-  const int max_critical_stack_depth = 10;
-  int depth = 0;
+  bool agents_loaded = JvmtiAgentList::has_agents();
```

Comment:
I see that for dynamically loaded agents we add to the list after loading the agent. Maybe we should check `JvmtiEnvBase::environments_might_exist()`?